### PR TITLE
Fix some bugs in `SplitContainer`

### LINKED ISSUE #43749: Custom dock with minimal size appears bigger than necessary
**Godot version:**

3.2.4 beta2

**Issue description:**

uyBuazcFLC
The dock has minimum horizontal size. It looks like the panel is resized to this size, but it's also added to the final size, so the left panel is wider than necessary. But when you manually shrink the width to minimal, when the dock is switched out the additional size is still subtracted resulting in tiny panel.

**Minimal reproduction project:**

[ReproductionProject.zip](https://github.com/godotengine/godot/files/5577793/ReproductionProject.zip)

### LINKED ISSUE #64273: Import dock randomly causes sidebar size jumps
### Godot version

e9e9e92

### System information

Windows 10 x64

### Issue description

Seems like minimum size of the import dock varies greatly and it can abruptly change sidebar's width. It always happens when switching to Import tab, but also randomly on it's own.
ezgif com-gif-maker

Might be caused #59303
But the random jumps and minsize weirdness looks similar to #43749

### Steps to reproduce

1. Select different files in the filesystem dock (scene, texture, sample etc)
2. The left dock will randomly change size
3. If not, switch to Import tab and observe size jump

### Minimal reproduction project

_No response_

### LINKED ISSUE #64851: Some Controls in the Editor reset to their minimum size after restart
### Godot version

v4.0.alpha.custom_build [d0a2a4c98]

### System information

Windows 10

### Issue description

Typically, I recall the editor remembering each panel's proportions, but this no longer happens consistently anymore.

For example, after increasing the Script List's size and restarting the editor, it keeps getting set to its minimum size. This is not the only time this occurs, but it's the most consistent I've been able to document.
image

### Steps to reproduce

Resize the Script List in the Script Editor, Restart the Editor, look at the Script List again.

### Minimal reproduction project

No need

### LINKED ISSUE #64866: Editor: Right sidebar gets priority when resizing
### Godot version

6e390fa9abf2101f8772c4ef6694471153105636

### System information

Fedora Silverblue 36, Ryzen 2400G

### Issue description

The right sidebar gets priority over the main content when resizing the window, and the initial size barely makes a difference:

[Screencast from 25-08-22 05:40:20.webm](https://user-images.githubusercontent.com/347965/186569688-1c5b2807-95bd-4870-be4c-a70abeb413d3.webm)

Doesn't happen with the left sidebar, and it must have been a very recent change, AFAIK my last build from 1-2 days ago did not have this problem. #64800 may be related @KoBeWi 

### Steps to reproduce

1. Resize window? =)

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by Rindbee**
According to the expand flag of the child controls, the position of the origin.

| expand flag |             1st=`false`       |                  1st=`true`        |
| :---------: | :---------------------------: | :--------------------------------: |
| 2nd=`false` | The end of 1st's minimum rect | The begin of 2nd's minimum rect    |
| 2nd=`true`  | Same as above                 | The middle of the container's rect |

`split_offset_origin`/`prev_split_offset_origin` will record the  position of the origin of `split_offset`, and their sign show which side of the container the origin is relative to.

~~Fix #64851, fix #64866.~~
Fix #43749, fix #64273.

Can replace the role of `use_hidden_tabs_for_min_size` (introduced by #25353) when the `TabContainer` has a `SplitContainer` type ancestor. This will prevent #64273 from appearing.

### CODE REVIEW

**Comment by KoBeWi** [comment]
Wow, it fixes the regressions while keeping the original issue fixed. Good job.

Needs a check from @groud

**Comment by groud** [comment]
According to the comments in the code, I think there's a misconception about what should be the "zero". Let me list the situations:
- If no child is expanded or if the right child only is expanded: the zero is at the end of the first child's minimum rect,
- If the left child only is expanded: the zero is at the start of the second child's minimum rect,
- If both are expanded: the zero is in the position relative to each child expand_ratio. So, for example, if the first child has an expand ratio of 2 and the second one an expand ratio of 1, then the zero should be at 2/3 of the full width. By default expand ratios are set to one, which means it's the middle.

Can you confirm the last case is processed correctly ?

**Comment by Rindbee** [comment]
"Zero" is expected to be the case where expand is enabled for both controls, otherwise, it will not be `0`. But it seems to be broken when the minimum size of the child control without expand is `0`. 

https://github.com/godotengine/godot/blob/ac7b5111f548e87b3555afb4f2a83ecd1ffbe85e/doc/classes/SplitContainer.xml#L21-L23

By the way. What "collapsed" means, it currently only seems to disable the grabber. When expand is enabled for both child controls, it also divides the space proportionally.

**Comment by groud** [comment]
> "Zero" is expected to be the case where expand is enabled for both controls, otherwise, it will not be 0. But it seems to be broken when the minimum size of the child control without expand is 0.

Not sure what you mean, but the behavior I described in my message is when the `split_offset`value is 0. This should be the default behavior.

> By the way. What "collapsed" means, it currently only seems to disable the grabber. When expand is enabled for both child controls, it also divides the space proportionally.

Yeah that makes sense. It's just a way to "reset" the splitting state and block it. Basically like if you would set the `split_offset` value to 0 and disable the grabber.

**Comment by Rindbee** [comment]
> > "Zero" is expected to be the case where expand is enabled for both controls, otherwise, it will not be 0. But it seems to be broken when the minimum size of the child control without expand is 0.
> 
> Not sure what you mean, but the behavior I described in my message is when the `split_offset`value is 0. This should be the default behavior.

That is the description of `no_offset_middle_sep`/`prev_no_offset_middle_sep`, not for `split_offset`. They are used to record the minimum size and position (left or right) of child controls.

For `split_offet`, the more complicated situation may be that the minimum size of the child control changes: it seems that `split_offet` needs to be clamped when the minimum size becomes larger.

**Comment by groud** [comment]
> That is the description of no_offset_middle_sep/prev_no_offset_middle_sep, not for split_offset. They are used to record the minimum size and position (left or right) of child controls.

I mean. It's in their name: `no_offset_middle_sep` is the position of the seperation when `split_offset` is `zero`. So all I am asking is whether the behavior I described above is still here after this PR. 

Split offset is a value that gets added to `no_offset_middle_sep` to get the final split position. And indeed, it gets clamped to respected each child minimum size.

**Comment by Rindbee** [comment]
`split_offset` is consistent with the previous behavior.

The meaning of `no_offset_middle_sep` has changed.
In simple terms `no_offset_middle_sep` now indicates the relative position of the origin (single axis).

It is also necessary to calculate the absolute position (relative to the left/top side of the container, not the first child control) of the origin (the **current** one and the **previous** one) by using the **current size** of the container.

In this way, the grabber can be positioned relative to the second child control, to keep the second child control‘s size fixed. Especially when the size of the container changes.

Expand only happens when the child control's expand is enabled. This is what this pr expects, but considering the legacy code, if neither is enabled, the expand effect takes effect on the second child control.

**Comment by Rindbee** [comment]
`split_offset` is consistent with the previous behavior.

The meaning of `no_offset_middle_sep` has changed.
In simple terms `no_offset_middle_sep` now indicates the relative position of the origin (single axis).

It is also necessary to calculate the absolute position (relative to the left/top side of the container, not the first child control) of the origin (the **current** one and the **previous** one) by using the **current size** of the container.

In this way, the grabber can be positioned relative to the second child control, to keep the second child control‘s size fixed. Especially when the size of the container changes.

Expand only happens when the child control's expand is enabled. This is what this pr expects, but considering the legacy code, if neither is enabled, the expand effect takes effect on the second child control.

**Comment by groud** [comment]
> In simple terms no_offset_middle_sep now indicates the relative position of the origin (single axis).

Then it should be renamed.

> Expand only happens when the child control's expand is enabled. This is what this pr expects, but considering the legacy code, if neither is enabled, the expand effect takes effect on the second child control.

Yeah, that makes sense. That's what I described as the expected behavior.

Anyway, I can't really tell about this PR. For me it does not look right as I think there should not really have the need for a check on a value that should be "linear". Maybe that's just a matter of renaming things though. If @KoBeWi thinks this PR is right, I trust him for the merge, but if others issues arise, I think we shall better revert everything.

**Comment by KoBeWi** [comment]
> If @KoBeWi thinks this PR is right, I trust him for the merge, but if others issues arise, I think we shall better revert everything.

Yeah I just tested all related issues and all seems fine. Can't say if nothing new will appear though.

**Comment by akien-mga** [comment]
I think it would be cleaner to revert the two previous PRs first, and then rebase this one to merge it. I can push that now.
It's hard to follow the logic and debug issues when so many regression fixes are piled on top of each other.

**Comment by akien-mga** [comment]
> I think it would be cleaner to revert the two previous PRs first, and then rebase this one to merge it. I can push that now. It's hard to follow the logic and debug issues when so many regression fixes are piled on top of each other.

Done with ecfff5b75e175a34cf891b4e0547ab761f212d44. This PR can be rebased to include only the new changes that would fix #43749 again without introducing the regressions found so far.

**Comment by Rindbee** [comment]
#43749 is actually caused by the change of the minimum size of the child controls.

There, we want the grabber to be fixed to one side of the container. To implement it we need the following information:
1. Which side needs to be fixed?
2. Grabber offset relative to that side.

The record of the previous information now only has two variables, `prev_no_offset_middle_sep` and `split_offset`, although the position of the grabber relative to the **left** side of the container can be calculated. But they can't say which side they want to be fixed to before. 

So this pr directly uses `prev_no_offset_middle_sep` to record the origin information of the previous `split_offset`. It also shows the previous state of expand (via `1`/`0`/`-1`), which side we want to be fixed to.

`prev_no_offset_middle_sep` adds `split_offset` to calculate the offset of the grabber relative to the side of the container that you want to fix.

**Comment by Rindbee** [comment]
Superseded by #65028.
